### ISSUE TITLE
vector_sort() could be much faster

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2286]
It is well-known that functions that use a non-inlineable custom comparator, such as `qsort()` in the C standard library, perform much worse than type-specialized sorting functions. It appears that this is not mitigated even by compiling igraph with LTO (see `igraph_qsort()` benchmark). Thus there is an opportunity for a very significant speedup for functions such as `igraph_vector_sort()`, `igraph_vector_int_sort()`, etc.

This is not a high priority because these functions are only used in a few places in the library. The most performance critical sorting operation is the internal sorting of the edge list when constructing a graph. This does not use `igraph_qsort()`. It uses a specialized counting sort instead.

#### References:

 - https://theory.stanford.edu/~amitp/rants/c++-vs-c/
 - https://travisdowns.github.io/blog/2019/05/22/sorting.html

**Comment by ntamas** [comment]
If we decide to mess with the sorting algorithm, shouldn't we consider using [`timsort`][1] or some other alternative sort algorithm that has been being used in other programming languages? (timsort is behind the sort implementation of Python, Java, Android, GNU Octave, Swift etc).

[1]: https://en.wikipedia.org/wiki/Timsort

**Comment by szhorvat** [comment]
Yes, alternative sorting algorithms are also interesting to look into. I suggest discussing those through a separate issue.

What I had in mind when opening this issue was providing specialized versions of the sorting function for reals and integers that properly inline the comparator. This idea applies to any sorting algorithm, including timsort.

However, integer and real versions of `igraph_qsort()` are not used in any performance critical part of igraph. In fact, they are barely used at all. The sorting for the core graph data structure is done using counting sort through a specialized function.

Thus this is low priority.
